compiler/rustc_codegen_llvm/src/builder/gpu_offload.rs

```diff
@@ -477,8 +470,27 @@ pub(crate) fn gen_call_handling<'ll, 'tcx>(
     let mut geps = vec![];
     let i32_0 = cx.get_const_i32(0);
     for &v in args {
-        let gep = builder.inbounds_gep(cx.type_f32(), v, &[i32_0]);
-        vals.push(v);
+        let name = String::from_utf8(get_value_name(v)).unwrap();
+        let (base_val, gep_base) = match cx.type_kind(cx.val_ty(v)) {
+            TypeKind::Pointer => (v, v),
+            _ => {
+                let addr =
+                    builder.direct_alloca(cx.val_ty(v), Align::EIGHT, &format!("{}.addr", name));
+                let casted =
+                    builder.direct_alloca(cx.type_i64(), Align::EIGHT, &format!("{}.casted", name));
+                builder.store(v, addr, Align::EIGHT);
+
+                let loaded = builder.load(cx.val_ty(v), addr, Align::EIGHT);
+                builder.store(loaded, casted, Align::EIGHT);
+
+                let casted_val = builder.load(cx.type_i64(), casted, Align::EIGHT);
+                (casted_val, casted)
+            }
+        };
+
+        let gep = builder.inbounds_gep(cx.type_f32(), gep_base, &[i32_0]);
```

Comment:
The idea is generally to move away from typed geps here, e.g.
https://www.npopov.com/2025/01/05/This-year-in-LLVM-2024.html
https://github.com/llvm/llvm-project/pull/68882

Since we only index at 0 it doesn't differ (4*0=0), but can you change it from f32 to i8 to not confuse other readers about the meaning?